File: Algorithm/DynamicProgramming/atCoder/A23/A23.php
<?php

fscanf(STDIN, "%d %d", $N, $M);

// クーポンごとのビットマスクを作成
$coupons = [];
for ($i = 0; $i < $M; $i++) {
    $line = trim(fgets(STDIN));
    $items = array_map('intval', explode(" ", $line));
    $bit = 0;
    for ($j = 0; $j < $N; $j++) {
        if ($items[$j] == 1) {
            $bit |= (1 << $j);
        }
    }
    $coupons[] = $bit;
}

$goal = (1 << $N) - 1; // すべての品目が揃った状態
$visited = array_fill(0, 1 << $N, false);
$queue = new SplQueue();
$queue->enqueue([0, 0]); // [現在の状態, 使用したクーポン数]
$visited[0] = true;

while (!$queue->isEmpty()) {
    list($state, $count) = $queue->dequeue();

    if ($state === $goal) {
        echo $count . PHP_EOL;
        exit;
    }

    foreach ($coupons as $coupon) {
        $next = $state | $coupon;
        if (!$visited[$next]) {
            $visited[$next] = true;
            $queue->enqueue([$next, $count + 1]);
        }
    }
}

echo -1 . PHP_EOL;

File: Algorithm/DynamicProgramming/atCoder/A23/A23.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type State struct {
	bitmask int
	steps   int
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	firstLine := strings.Fields(scanner.Text())
	N, _ := strconv.Atoi(firstLine[0])
	M, _ := strconv.Atoi(firstLine[1])

	coupons := make([]int, M)

	for i := 0; i < M; i++ {
		scanner.Scan()
		line := strings.Fields(scanner.Text())
		bit := 0
		for j := 0; j < N; j++ {
			val, _ := strconv.Atoi(line[j])
			if val == 1 {
				bit |= 1 << j
			}
		}
		coupons[i] = bit
	}

	goal := (1 << N) - 1
	visited := make([]bool, 1<<N)

	queue := []State{{bitmask: 0, steps: 0}}
	visited[0] = true

	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]

		if current.bitmask == goal {
			fmt.Println(current.steps)
			return
		}

		for _, coupon := range coupons {
			next := current.bitmask | coupon
			if !visited[next] {
				visited[next] = true
				queue = append(queue, State{bitmask: next, steps: current.steps + 1})
			}
		}
	}

	fmt.Println(-1)
}
